src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -329,6 +280,204 @@ void runtime::gpu::CudaKernelBuilder::get_topk(codegen::CodeWriter& writer,
+                uint32_t unroll_num = 8;
+                writer << "for(; idx" << last_r_idx << " + " << unroll_num << " - 1 < reduce_shape"
+                       << last_r_idx << "; idx" << last_r_idx << " += " << unroll_num << ")\n";
+                writer.block_begin();
+                {
+                    for (int k = 0; k < unroll_num; k++)
+                    {
+                        writer << "input_i = expf(in[reduce_idx] - r_max);\n";
+                        writer << "y = input_i - c;\n";
+                        writer << "t = r_sum + y;\n";
+                        writer << "c = (t - r_sum) - y;\n";
+                        writer << "r_sum = t;\n";
+                        writer << "reduce_idx += step;\n";
+                    }
+                }
+                writer.block_end();
+                writer << "for(; idx" << last_r_idx << " < reduce_shape" << last_r_idx << "; idx"
+                       << last_r_idx << "++)\n";
+                writer.block_begin();
+                {
+                    writer << "input_i = expf(in[reduce_idx] - r_max);\n";
+                    writer << "y = input_i - c;\n";
+                    writer << "t = r_sum + y;\n";
+                    writer << "c = (t - r_sum) - y;\n";
+                    writer << "r_sum = t;\n";
+                    writer << "reduce_idx += step;\n";
+                }
+                writer.block_end();
+            }
+            for (int64_t j = 0; j < last_r_idx; j++)
+            {
+                writer.block_end();
+            }
+            writer.block_end();
+
+            //divide
+            writer.block_begin();
+            for (int64_t j = 0; j < last_r_idx; j++)
+            {
+                writer << "for(int idx" << j << " = 0; idx" << j << "< reduce_shape" << j << "; idx"
+                       << j << "++)\n";
+                writer.block_begin();
+            }
+            {
+                writer << "uint32_t reduce_idx = in_idx;\n";
+                for (int64_t j = 0; j < last_r_idx; j++)
+                {
+                    writer << "reduce_idx += idx" << j << " * reduce_strides" << j << ";\n";
+                }
+                writer << "uint32_t step = reduce_strides" << last_r_idx << ";\n";
+                writer << "int idx" << last_r_idx << " = 0;\n";
+                // unroll last reduction axis
+                uint32_t unroll_num = 8;
+                writer << "for(; idx" << last_r_idx << " + " << unroll_num << " - 1 < reduce_shape"
+                       << last_r_idx << "; idx" << last_r_idx << " += " << unroll_num << ")\n";
+                writer.block_begin();
+                {
+                    for (int k = 0; k < unroll_num; k++)
+                    {
+                        writer << "input_i = expf(in[reduce_idx] - r_max) / r_sum;\n";
+                        writer << "out[reduce_idx] = input_i;\n";
+                        //writer << "out[reduce_idx] /= r_sum;\n";
```

Comment:
Writes should be non-blocking and so loading the pre-exponentiated value would work, but it seems slightly slower in my tests. Remove commented line.